### ISSUE TITLE
What's the correct atomFamily cleanup pattern?

### ISSUE THREAD

**Issue by LucasEThomas** [facebookexperimental/Recoil#2166]
tldr; the docs seem to be missing instructions related to removing/resetting atomFamily state when a key gets removed.

I must display a list of item components with some state in each item. Items can be added, removed, and their order rearranged. To do this I have an atom called Items that contains the keys for each Item, and I have an atomFamily that is indexed by those keys for the contents of each item (such as label).

This works nicely. My troubles begin when I want to remove an item. I do this by removing the item's key from the Items array. This causes the ItemComponent to be removed from the page, but it does not cause any of the related atomFamily atoms to be removed. If I add a new item later on with the same key, that atom will instantiate with the residual data from the deleted one.

I can't seem to find a well documented / accepted pattern for cleaning up these atoms in the docs or elsewhere on the internet. 

One thing I tried was reseting the atomFamily atoms in the ItemComponent's useEffect cleanup function:

``` typescript
const ItemComponent = ({key}) => { 
  const [label, setLabel] = useRecoilState(lineItemLabelState(key))
  const resetLabel = useResetRecoilState(lineItemLabelState(key))

  // reset recoil for this item on component cleanup
  useEffect(() => () => resetLabel(), [])
  // ....
}
```
This works great, but it introduces a new problem. This cleanup function also gets called when I rearrange Items, and thus resets the contents of the Items that got rearranged. So I could add some logic to check if those items' keys were really removed from the Items array, but this would be an O(n^2) time operation as more and more items get added. Also it just feels kind'a wrong.

Is there a better / more accepted pattern I'm overlooking? Perhaps this could be clarified in the docs because I imagine it's a common need.

**Comment by taylorcode** [comment]
There's a feature named `retainedBy` which hasn't been exposed in the Recoil documentation. This is a great solution to the cleanup problem, but unfortunately it does not appear to work yet: https://github.com/facebookexperimental/Recoil/issues/2173

**Comment by LucasEThomas** [comment]
Thanks for pointing this out! `retainedBy` looks like a perfect and elegant solution to my exact need. Too bad it's not working yet. I will be watching #2173 closely
